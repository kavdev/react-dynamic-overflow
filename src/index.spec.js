import React from "react";
import { shallow, mount } from "enzyme";
import DynamicOverflow from "./";

const defaultProps = {
  list: ({ tabRef }) => {
    return [
      <span
        key={0}
        ref={() => {
          tabRef({ clientWidth: 100 });
        }}
      >
        First tab
      </span>,
      <span key={1}>Second tab</span>,
      <span key={2}>Third tab</span>
    ];
  },
  throttle: 123
};

const events = {};

beforeEach(() => {
  window.addEventListener = jest.fn(
    (eventName, cb) => (events[eventName] = cb)
  );
  window.removeEventListener = jest.fn(
    eventName => (events[eventName] = undefined)
  );
});

afterEach(() => {
  window.addEventListener.mockClear();
  window.removeEventListener.mockClear();
});

it("should call children function with correct arguments when all elements can be visible", () => {
  const mockChildFunction = jest.fn(({ visibleElements, containerRef }) => (
    <div ref={() => containerRef({ clientWidth: 1000 })}>{visibleElements}</div>
  ));
  const wrapper = mount(
    <DynamicOverflow {...defaultProps}>{mockChildFunction}</DynamicOverflow>
  );

  expect(mockChildFunction.mock.calls).toMatchSnapshot();
});

it("should always call children function with one element", () => {
  const mockChildFunction = jest.fn(({ visibleElements, containerRef }) => (
    <div ref={() => containerRef({ clientWidth: 0 })}>{visibleElements}</div>
  ));
  const wrapper = mount(
    <DynamicOverflow {...defaultProps}>{mockChildFunction}</DynamicOverflow>
  );

  expect(mockChildFunction.mock.calls).toMatchSnapshot();
});

it("should call children function with correct arguments when some elements are hidden", () => {
  const mockChildFunction = jest.fn(({ visibleElements, containerRef }) => (
    <div ref={() => containerRef({ clientWidth: 300 })}>{visibleElements}</div>
  ));
  const wrapper = mount(
    <DynamicOverflow {...defaultProps}>{mockChildFunction}</DynamicOverflow>
  );

  expect(mockChildFunction.mock.calls).toMatchSnapshot();
});

it("should add event listener when mounting", () => {
  const mockChildFunction = jest.fn(({ visibleElements, containerRef }) => (
    <div ref={() => containerRef({ clientWidth: 300 })}>{visibleElements}</div>
  ));
  const wrapper = mount(
    <DynamicOverflow {...defaultProps}>{mockChildFunction}</DynamicOverflow>
  );

  const resizeEventListener = window.addEventListener.mock.calls.find(
    event => event[0] === "resize"
  );
  expect(resizeEventListener).toMatchSnapshot();
});

it("should remove event listener when unmounting", () => {
  const mockChildFunction = jest.fn(({ visibleElements, containerRef }) => (
    <div ref={() => containerRef({ clientWidth: 300 })}>{visibleElements}</div>
  ));
  const wrapper = mount(
    <DynamicOverflow {...defaultProps}>{mockChildFunction}</DynamicOverflow>
  );
  wrapper.unmount();

  const resizeEventListener = window.removeEventListener.mock.calls.find(
    event => event[0] === "resize"
  );
  expect(resizeEventListener).toMatchSnapshot();
});

it("should recalculate visible elements when resizing window", () => {
  const containerNode = { clientWidth: 1000 };
  let mockChildFunction = jest.fn(({ visibleElements, containerRef }) => (
    <div ref={() => containerRef(containerNode)}>{visibleElements}</div>
  ));
  let wrapper = mount(
    <DynamicOverflow {...defaultProps}>{mockChildFunction}</DynamicOverflow>
  );

  mockChildFunction.mockClear();

  containerNode.clientWidth = 0;
  events.resize();
  expect(mockChildFunction).toHaveBeenCalledTimes(1);
  expect(mockChildFunction.mock.calls[0]).toMatchSnapshot();
});
