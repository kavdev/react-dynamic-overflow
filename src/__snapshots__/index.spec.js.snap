// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add event listener when mounting 1`] = `
Array [
  "resize",
  [Function],
]
`;

exports[`should always call children function with one element 1`] = `
Array [
  Array [
    Object {
      "containerRef": [Function],
      "overflowElements": Array [],
      "visibleElements": Array [
        <span>
          First tab
        </span>,
        <span>
          Second tab
        </span>,
        <span>
          Third tab
        </span>,
      ],
    },
  ],
  Array [
    Object {
      "containerRef": [Function],
      "overflowElements": Array [
        <span>
          Second tab
        </span>,
        <span>
          Third tab
        </span>,
      ],
      "visibleElements": Array [
        <span>
          First tab
        </span>,
      ],
    },
  ],
]
`;

exports[`should call children function with correct arguments when all elements can be visible 1`] = `
Array [
  Array [
    Object {
      "containerRef": [Function],
      "overflowElements": Array [],
      "visibleElements": Array [
        <span>
          First tab
        </span>,
        <span>
          Second tab
        </span>,
        <span>
          Third tab
        </span>,
      ],
    },
  ],
  Array [
    Object {
      "containerRef": [Function],
      "overflowElements": Array [],
      "visibleElements": Array [
        <span>
          First tab
        </span>,
        <span>
          Second tab
        </span>,
        <span>
          Third tab
        </span>,
      ],
    },
  ],
]
`;

exports[`should call children function with correct arguments when some elements are hidden 1`] = `
Array [
  Array [
    Object {
      "containerRef": [Function],
      "overflowElements": Array [],
      "visibleElements": Array [
        <span>
          First tab
        </span>,
        <span>
          Second tab
        </span>,
        <span>
          Third tab
        </span>,
      ],
    },
  ],
  Array [
    Object {
      "containerRef": [Function],
      "overflowElements": Array [
        <span>
          Third tab
        </span>,
      ],
      "visibleElements": Array [
        <span>
          First tab
        </span>,
        <span>
          Second tab
        </span>,
      ],
    },
  ],
]
`;

exports[`should recalculate visible elements when resizing window 1`] = `
Array [
  Object {
    "containerRef": [Function],
    "overflowElements": Array [
      <span>
        Second tab
      </span>,
      <span>
        Third tab
      </span>,
    ],
    "visibleElements": Array [
      <span>
        First tab
      </span>,
    ],
  },
]
`;

exports[`should remove event listener when unmounting 1`] = `
Array [
  "resize",
  [Function],
]
`;
